Custom Feedback Paths: Frontend

JosephineKwa opened the pull request:
@damithc changes made:
1. Added handsontable library for the spreadsheet functionality
2. Added an editable spreadsheet to display the individual giver to recipient feedback paths.
3. Added a simple UI test for the showing and hiding of the spreadsheet.

After clicking edit, the link to show/hide spreadsheet appears
![fp1](https://cloud.githubusercontent.com/assets/9999077/16517930/ba55f7f0-3fb2-11e6-956d-dc13f72db158.PNG)

After clicking "Show details and further customizations", the spreadsheet appears
![fp2](https://cloud.githubusercontent.com/assets/9999077/16517932/bc37ec22-3fb2-11e6-9eff-261bf468295f.PNG)

Upon selecting a different giver (Session Creator), the spreadsheet is updated to reflect the new feedback paths
![fp3](https://cloud.githubusercontent.com/assets/9999077/16517971/e4810736-3fb2-11e6-8477-b52c5c2a30a9.PNG)

If the recipient type is invalid for the chosen giver type (or if no recipients are available for the giver), only the givers are displayed
![fp4](https://cloud.githubusercontent.com/assets/9999077/16518034/52b5b36e-3fb3-11e6-80c2-5c1856c85f4a.PNG)

If the data in the spreadsheet is changed, the giver type and recipient type options will be changed to "Custom"
![fp5](https://cloud.githubusercontent.com/assets/9999077/16518145/d4c459aa-3fb3-11e6-92e4-9a8fa7637b22.PNG)

Some notes:
1. Selecting "Nobody specific" as the recipient type will result in the recipients being "%GENERAL%"
2. Selecting "Instructors in this course" as the giver type and "Giver's team" as the recipient type will result in the recipients being "Instructors"
3. Columns and rows can be resized
4. When the the giver and recipient types are custom, all the visibility rows are shown

wkurniawan07 commented:
Not going to comment on the PR itself (unless arrowed later on):

If you're not merging to `master`, the Travis CI build will not be triggered. It will be hard to maintain a check on whether things break or not. I think you should open a PR from the `custom-feedback-paths` branch so that every time it gets updated (from the smaller PRs) a CI build is triggered.
Your small PRs cannot be checked prior to getting merged to the `custom-feedback-paths` branch, but I guess that's one thing that can't be helped. If things go wrong you can do a hotfix in the `custom-feedback-paths` branch itself.

damithc commented:
Columns can be named to match the column contents? e.g. Feedback giver, Feedback recipient
Is there going to some validation? Can be done later though.
Will the cells act as dropdowns too (i.e. select a value from possible values)?. Again, can be done as a separate PR.
Is the library free for anything, or non-commercial only?

JosephineKwa commented:
1. Should be able rename the column header to match the column content.
2. There will be both frontend and backend validation. I intended for it to be done later as this PR is mainly to get the general UI up.
3. Should be possible to have the cell act as dropdowns too.
4. It is free for anything as it is [under the MIT Licence](https://docs.handsontable.com/0.26.0/tutorial-licensing.html).

JosephineKwa commented:
After discussion with @damithc ,
Column headers changed to Feedback giver and Feedback recipient, spreadsheet is read-only for predefined feedback participant types
![fp6](https://cloud.githubusercontent.com/assets/9999077/16641618/2b649042-4435-11e6-8289-1565c159a7f8.PNG)

Upon clicking "Customize further", the spreadsheet becomes editable and the feedback participant type dropdowns switch to "Custom". The user can select the individual feedback participant from a searchable dropdown in each cell
![fp7](https://cloud.githubusercontent.com/assets/9999077/16641620/2d1f4a4e-4435-11e6-9b4c-bcafdf727ed4.PNG)

If an invalid feedback participant is input into the cell, the cell will be highlighted red
![fp8](https://cloud.githubusercontent.com/assets/9999077/16641621/2f13505c-4435-11e6-901d-601c6886e134.PNG)

JosephineKwa commented:
Also increased height of spreadsheet due to display issues for the cell dropdown as a result of the spreadsheet being too short.

@YongJieYongJie ready for review!

YongJieYongJie commented:
@JosephineKwa 
Here's some issues and suggestions:
 - the first option is automatically selected when clicking away from the dropdown menu, the may be unexpected for the user
 - perhaps put the "Customize Further" button below the "How to use" text, and refer to the button in the text itself (e.g., "the spreadsheet to the right shows the current feedback path according to the chosen options, you can fully customize the paths by clicking on the button below")
 - perhaps just name the button "Customize" instead of "Customize Further" because it is not really "Further"
 - Maybe the "Add rows" feature can be something like Google sheets (with a certain number pre-filled and less repetition of the word "row"):

![add rows](https://cloud.githubusercontent.com/assets/14101781/16645284/7251f4fc-4455-11e6-9535-9d56dc94d514.png)


I think this is most of the things of now. When we combine this with #5604, we will have to work on the UI again.

JosephineKwa commented:
I will be removing the add rows functionality as it doesn't serve much purpose. This is because there will always be an extra row at the bottom (specified as an option when initialising the spreadsheet), and pasting data into the spreadsheet, where the rows in the data exceeds the number or rows in the spreadsheet, will result in the auto-resizing of the spreadsheet, i.e. rows are automatically added to accommodate the data.

damithc commented:
If pasting is the only way to add new rows, may be better to keep the 'add rows' function. It's a bit redundant, but some people may want to add blank rows first before entering details.

JosephineKwa commented:
I've made updates based on the comments given, 
1. Renamed the customize button and referenced the button in the instructions
2. Added a blank option to the dropdown to address the issue with clicking away from the dropdown
3. Edited the appearance of the add rows feature
![fp12](https://cloud.githubusercontent.com/assets/9999077/16649329/0b71ddea-446b-11e6-96df-4831f8a35dcf.PNG)

YongJieYongJie commented on src/main/java/teammates/common/datatransfer/FeedbackParticipantType.java:
```diff
@@ -97,6 +98,8 @@ public String toSingularFormString() {
         case OWN_TEAM:
             return "team";
+        case CUSTOM:
+            return "custom feedback participant";
```
What are the places that this will be shown, perhaps we should change the whole message rather than just use "custom feedback participant"

JosephineKwa replied:
This is used in places where the visibility message is displayed, like the feedbackEdit and feedbackSubmissionEdit pages. This will have to be changed, however I would prefer to do it later, perhaps when I am working on the access control aspect of this feature.

YongJieYongJie commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditPageData.java:
```diff
@@ -293,4 +314,21 @@ public boolean getHasError() {
+    }
+    
+    public String getInstructorsDataAsString() {
+        return Sanitizer.sanitizeForHtml(Sanitizer.sanitizeListForCsv(instructorsData).toString());
```
Is it possible to use the list version of `sanitizeForHtml()`?

YongJieYongJie commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditPageData.java:
```diff
@@ -293,4 +314,21 @@ public boolean getHasError() {
+    
+    public String getStudentsDataAsString() {
+        JSONObject json = new JSONObject(studentsData);
+        return Sanitizer.sanitizeForHtml(json.toString());
```
Are we converting the `studentData` to JSON so that it can be easily converted to string? If so, maybe leave a comments so future readers will not be confused by the JSON conversion.

YongJieYongJie commented on src/main/webapp/WEB-INF/tags/instructor/feedbackEdit/customFeedbackPathsData.tag:
```diff
@@ -0,0 +1,7 @@
+
+<div id="session-creator-data" data-session-creator="${sessionCreatorData}"></div>
+<div id="students-data" data-students="${studentsData}"></div>
+<div id="instructors-data" data-instructors="${instructorsData}"></div>
```
_Note: Actually instructor and student data are already being passed to the preview session form; unfortunately, they are wrapped in inaccessible options tag =(_

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -863,3 +877,41 @@ function getQuestionIdSuffix(questionNum) {
     return idSuffix;
 }
 
+function toggleCustomFeedbackPathsDisplay(toggleLink) {
```
Can we put the feedback path related methods in `instructorEditPage/feedbackPath.js`?
The methods are separated into the said file in #5744, I think should do the same here so when merging these two in the future the feedback path related methods will not be split in two locations.

JosephineKwa replied:
I think I would prefer to do it at a later stage, at least after the PR has been merged in (I will be integrating master back into the `custom-feedback-paths` branch periodically), so that it resolving conflicts will be (hopefully) less messy.

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -863,3 +877,41 @@ function getQuestionIdSuffix(questionNum) {
+    });
+}
+
+function removeUnselectedCustomOptions() {
```
Can we remove these in the `InstructorFeedbackEditPageData` when generating the options tag? Somewhere near `getParticipantOptions()`

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit.js:
```diff
@@ -646,16 +657,19 @@ function copyOptions() {
-    $currGiver.val($prevGiver.val());
+    // If previous feedback path was custom, there's no need to copy
+    if ($prevGiver.val() === 'CUSTOM' || $prevRecipient.val() === 'CUSTOM') {
+        return;
```
Maybe for consistency's sake, we should copy the custom feedback path too; otherwise, I would recommend adding a prominent note to the header comments regarding this special case

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+}
+
+function enableAllVisibilityOptionsRowsForQuestion($questionForm) {
+    enableRow($questionForm.children(':first'), 1);
```
Maybe we can use the `enableAllRows()` in `visibilityOptions.js` in #5803. Since it is not merged yet, can just copy the function to `visibilityOptions.js`

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+$(document).ready(function() {
+    initialiseCustomFeedbackPathsData();
+    initialiseFeedbackPathsSpreadsheets();
+    bindEventHandlers();
```
Since this is a whole new file, do you think we can namespace it. That way, we can have short method names without fear of populating global namespace.
E.g. this `bindEventHandlers()` may possibly confilct

More importantly, when adding methods, we'll consider whether it really belongs here. Also, when we eventually decide to revamp our js, I think a namespaced file is easier to convert

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+    bindEventHandlers();
+});
+
+function initialiseCustomFeedbackPathsData() {
```
Maybe add a header comment to describe briefly what is being done here

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+    teamNames = [];
+    
+    for (var studentEmail in studentEmailToTeamNameMap) {
+        if (studentEmailToTeamNameMap.hasOwnProperty(studentEmail)) {
```
Just to confirm, `studentEmail` is from `studentEmailToTeamNameMap` and we are also checking if `studentEmailToTeamNameMap` has a direct property as contained in `studentEmail`?

JosephineKwa replied:
Calling `hasOwnProperty` when iterating over a javascript object is to prevent also looping through the primitive object's prototype, although this is mainly necessary only when the object has prototype methods. Source: [This stackoverflow post](http://stackoverflow.com/questions/921789/how-to-loop-through-javascript-object-literal-with-objects-as-members). This check is also enforced by eslint.

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+        }
+    }
+    
+    allPossibleFeedbackGivers = [''];
```
Is this the workaround for the clicking away issue? If so, maybe add a comment to explain

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+        }
+    }
+    allPossibleFeedbackGivers = allPossibleFeedbackGivers.concat(teamNames);
+    allPossibleFeedbackRecipients = allPossibleFeedbackGivers.slice();
```
Just to confirm, `slice()` is used to create a deep copy?

JosephineKwa replied:
Yup `slice()` will create a deep copy of the array as it contains only primitives. If the array contained objects, it will create a shallow copy. (However in this case, as discussed, it will not matter if it is deep or shallow as `allPossibleFeedbackRecipients` and `allPossibleFeedbackGivers` are not meant to be changed after they have been initialised)

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+    allPossibleFeedbackGivers = allPossibleFeedbackGivers.concat(teamNames);
+    allPossibleFeedbackRecipients = allPossibleFeedbackGivers.slice();
+    allPossibleFeedbackRecipients.push(TEAM_NAME_INSTRUCTORS);
+    allPossibleFeedbackRecipients.push('%GENERAL%');
```
Instead of `%GENERAL%`, perhaps we want to use a more user-friendly string and change it back to `%GENERAL%` on submission

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+    allPossibleFeedbackRecipients.push('%GENERAL%');
+}
+
+function initialiseFeedbackPathsSpreadsheets() {
```
I think our codebase uses more of the American spelling, so `initialize...`

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+        var $questionForm = $(this);
+        var giverType = $questionForm.find('select[id^="' + FEEDBACK_QUESTION_GIVERTYPE + '"]').val();
+        var recipientType = $questionForm.find('select[id^="' + FEEDBACK_QUESTION_RECIPIENTTYPE + '"]').val();
+        generateFeedbackPathsSpreadsheet($questionForm.find('.custom-feedback-paths-spreadsheet'), giverType, recipientType);
```
Suggestion (not sure if better): perhaps just pass in the `$questionForm` to `generateFeedbackPathsSpreadsheet()` and let it find the container etc.

This is sort of the approach I took in #5671, where all methods refer to the `$questionForm`, and they know how to locate the elements concerned. Essentially, `$questionForm` serves the purpose of `questionNumber` in our .java files

JosephineKwa replied:
I passed in the `$questionForm.find('.custom-feedback-paths-spreadsheet')` object as it is the actual `$container` object that is used to init the handsontable. I thought this would be more appropriate for the `generateFeedbackPathsSpreadsheet` method as it specifically means "generating the feedback paths spreadsheet/handsontable" to this `$container`

Although, at this point I'm not sure which approach would be better, I might find that one way may be more suited for my purposes as development of this feature progresses.

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+                || recipientType === FEEDBACK_PARTICIPANT_TYPE_CUSTOM) {
+            $giverSelect.find('option[value="' + FEEDBACK_PARTICIPANT_TYPE_CUSTOM + '"]').remove();
+            $recipientSelect.find('option[value="' + FEEDBACK_PARTICIPANT_TYPE_CUSTOM + '"]').remove();
+        }
```
Is this actually testing for when user change from custom back to predefined? If so, maybe use if `giverType != ... ...`

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+        }
+        
+        giverType = $giverSelect.val();
+        recipientType = $recipientSelect.val();
```
Is this just using the default value after the custom option is removed? Should be set a standard default?

JosephineKwa replied:
Yup it uses the default value, generally this would be the first non-disabled option. If I use a standard default, I would need to consider the case where the question is a Team Contrib question as it only allows students to giver's team members and giver as the predefined feedback path. This might make the logic more complicated than necessary.

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+        var hotInstance = $container.handsontable('getInstance');
+        if (numRowsToAdd > 0) {
+            hotInstance.alter('insert_row', hotInstance.countRows(), numRowsToAdd);
+        }
```
Do you think we should extract a function here? Or will it be too messy

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,358 @@
+                $(this).append($customFeedbackParticipantTypeOption)
+                       .val(FEEDBACK_PARTICIPANT_TYPE_CUSTOM);
+            }
+        });
```
should try to keep the statements at similar level of abstraction

YongJieYongJie commented on src/test/java/teammates/test/driver/HtmlHelper.java:
```diff
@@ -403,6 +404,8 @@ private static String replaceUnpredictableValuesWithPlaceholders(String content)
                       .replaceAll("commentBar-" + REGEX_COMMENT_ID, "commentBar-\\${comment\\.id}")
                       .replaceAll("plainCommentText-" + REGEX_COMMENT_ID, "plainCommentText-\\${comment\\.id}")
+                      // randomly generated handsontable id
+                      .replaceAll("id=\"ht_" + REGEX_HANDSONTABLE_ID_SUFFIX + "\"", "id=\"\\${handsontable\\.id}\"")
```
maybe put the `ht_` into the regex itself?

YongJieYongJie commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java:
```diff
@@ -93,6 +93,8 @@ private void testGeneralQuestionOperations() throws Exception {
         testEditQuestionLink();
         testEditQuestionAction();
+        
+        testToggleFeedbackPathDisplayLink();
```
Maybe standardize naming with the `clickToggleCustomFeedbackPath...` below and add the word `custom`

YongJieYongJie commented on src/test/resources/pages/instructorFeedbackAddSuccess.html:
```diff
@@ -5,6 +5,12 @@
+  </div>
+  <div data-students="{&quot;CFeedbackUiT.charlie.tmms@gmail.tmt&quot;:&quot;Team 2&quot;,&quot;CFeedbackUiT.danny.tmms@gmail.tmt&quot;:&quot;Team 2&quot;}" id="students-data">
+  </div>
+  <div data-instructors="[&quot;teammates.helper@gmail.tmt&quot;, &quot;teammates.test1@gmail.tmt&quot;, &quot;teammates.test2@gmail.tmt&quot;]" id="instructors-data">
```
Hmm, since these are pure data anyway, perhaps we should just use the `<script>` and define javascript variables directly

Or maybe we can put these data attributes on the session/page related elements

YongJieYongJie commented:
@JosephineKwa Add (quite) some comments. Also added a lot of suggestions that I'm not sure if better, maybe can discuss more.

Additionally, I think the instructions on how to use can be improved too.

JosephineKwa commented:
@YongJieYongJie ready for review again!

YongJieYongJie commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditPageData.java:
```diff
@@ -195,8 +195,10 @@ private void buildNewQuestionForm(FeedbackSessionAttributes feedbackSession, int
                 
-                ElementTag optionTag = createOption(participantName, option.toString(), isSelected);
-                result.add(optionTag);
+                if (option != FeedbackParticipantType.CUSTOM || isSelected) {
```
maybe extract a boolean variable or use comment like `don't add option for custom feedback type unless it is already selected`

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -1,368 +1,379 @@
-var FEEDBACK_PARTICIPANT_TYPE_OWN_TEAM_MEMBERS_INCLUDING_SELF = 'OWN_TEAM_MEMBERS_INCLUDING_SELF';
-var FEEDBACK_PARTICIPANT_TYPE_NONE = 'NONE';
-var TEAM_NAME_INSTRUCTORS = 'Instructors';
+var customFeedbackPaths = {
```
maybe use uppercase C?

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -1,368 +1,379 @@
+                var studentEmailsList = customFeedbackPaths.teamNameToStudentEmailsMap[teamName];
+                if (studentEmailsList === undefined) {
+                    studentEmailsList = [];
+                }
```
Willl `studentEmailsList = studentEmailslist || []` work? If so, which is the more common way to initialize?

JosephineKwa replied:
Yup it works, I'm not sure which way is more common though

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -1,368 +1,379 @@
+$(document).ready(function() {
+    customFeedbackPaths.initializeCustomFeedbackPathsData();
+    customFeedbackPaths.initializeFeedbackPathsSpreadsheets();
+    customFeedbackPaths.bindEventHandlers();
```
as @wkurniawan07 suggested in https://github.com/TEAMMATES/teammates/pull/5744#discussion_r70209542, maybe put this in `instructorFeedbackEdit.js` so this `customFeedbackPaths` can be unit tested without needing `$(document).ready()`

YongJieYongJie commented on src/main/webapp/js/instructorFeedbackEdit/customFeedbackPaths.js:
```diff
@@ -0,0 +1,379 @@
+var customFeedbackPaths = {
```
maybe use uppercase C?

YongJieYongJie commented:
@JosephineKwa Added some comments! (fewer and lesser as compared to previous review)

JosephineKwa commented:
Ready for review again :)

YongJieYongJie commented:
@JosephineKwa Looks good! (other than the things you said to fix in subsequent parts)
